tree/forest/v7/inc/ROOT/RTree.hxx

```diff
@@ -0,0 +1,148 @@
+/// \file ROOT/RTree.hxx
+/// \ingroup Forest ROOT7
+/// \author Jakob Blomer <jblomer@cern.ch>
+/// \date 2018-10-04
+/// \warning This is part of the ROOT 7 prototype! It will change without notice. It might trigger earthquakes. Feedback
+/// is welcome!
+
+/*************************************************************************
+ * Copyright (C) 1995-2015, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#ifndef ROOT7_RTree
+#define ROOT7_RTree
+
+#include <ROOT/RStringView.hxx>
+#include <ROOT/RTreeUtil.hxx>
+#include <ROOT/RTreeView.hxx>
+
+#include <memory>
+#include <utility>
+
+namespace ROOT {
+namespace Experimental {
+
+class RTreeEntry;
+class RTreeModel;
+
+namespace Detail {
+class RTreeSink;
+class RTreeSource;
+}
+
+namespace Detail {
+
+// clang-format off
+/**
+\class ROOT::Experimental::RTree
+\ingroup Forest
+\brief The RTree represents a live dataset, whose structure is defined by an RTreeModel
+
+RTree connects the static information of the RTreeModel to a source or sink on physical storage.
+Reading and writing requires use of the corresponding derived class RInputTree or ROutputTree.
+RTree writes only complete entries (rows of the data set).  The entry itself is not kept within the
+RTree, which allows for multiple concurrent entries for the same RTree.  Besides reading an entire entry,
+the RTree can expose tree views that read only specific branches.
+*/
+// clang-format on
+class RTree {
+private:
+   /// All trees that use the same model share its ownership
+   std::shared_ptr<RTreeModel> fModel;
+   /// The number of entries is constant for reading and reflects the current number of Fill() operations
+   /// when writing
+   TreeIndex_t fNentries;
```

Comment:
I think this is an unnecessary obfuscation. I cannot tell if this is 32 or 64 bits, unsigned, signed, etc. Maybe a plain `size_t` would do?

Reply:
I think everything that denotes a number of entries cannot be `size_t` and friends, which would be architecture dependent. The `TreeIndex_t` is an integer long enough to hold the maximum possible number of entries in a tree or branch.